### ISSUE TITLE
Consider adding IAsyncDisposable support for Browser and Page

### ISSUE THREAD

**Issue by madelson** [hardkoded/puppeteer-sharp#1456]
Thanks for building and maintaining this library!

Today, `Browser` and `Page` have `CloseAsync()` + `Dispose()`.

Now that we have `IAsyncDisposable` and `await using` syntax, it would be nice for these types to also implement that interface and expose a `DisposeAsync()` method which just calls `CloseAsync()`.

This could be done while still targeting netstandard2.0 using https://www.nuget.org/packages/Microsoft.Bcl.AsyncInterfaces/.

If there is interest in this feature, I'd be happy to contribute it (if that would be helpful).

**Comment by kblok** [comment]
Go for it @madelson !

**Comment by madelson** [comment]
Great. I have some time set aside for this next week.

**Comment by madelson** [comment]
@kblok I've started working on this and I have a few thoughts/questions I thought would be worth running by you.

For `Browser`, we can implement `DisposeAsync` as `=> new ValueTask(CloseAsync())`. Because `Browser.CloseAsync()` caches the returned task, we can conveniently test this by simply asserting that `CloseAsync()` and `DisposeAsync().AsTask()` are the same reference. Does that seem like a reasonable testing approach to you or would you prefer to have a set of tests replicating whatever behavior we assert for `CloseAsync()`?

`Page` we can do the same implementation in theory for `Page`, but `Page.CloseAsync()` works a bit differently. Calling `CloseAsync()` the first time will hit:
```
                return Client.Connection.SendAsync("Target.closeTarget", new TargetCloseTargetRequest
                {
                    TargetId = Target.TargetId
                }).ContinueWith(task => Target.CloseTask);
```
I'm not sure what to make of the ContinueWith, since it is returning a `Task<Task>` but not calling `Unwrap` so we don't observe or wait on the inner `CloseTask`.

The second time `CloseAsync()` is called, we hit:
```
_logger.LogWarning("Protocol error: Connection closed. Most likely the page has been closed.");
return _closeCompletedTcs.Task;
```
This is giving back a different task whose purpose seems to be ensuring that (1) any exceptions hit by firing page close events are swallowed and (2) any continuations will run asynchronously. One strange thing about this is that consecutive calls to close can return tasks with (temporarily) inconsistent completion state.

I'm wondering whether it makes sense to adjust how `Page.CloseAsync()` works as part of this work. It seems like it would be safe to always return `_closeCompletedTcs.Task`. Another option would be to change the `Page` constructor so that the continuation it creates on `Target.CloseTask` (which fires the close events) is saved off. We could pass `TaskContinuationOptions.RunContinuationsAsynchronously` to that continuation to preserve the behavior we were getting from the TCS, and we could choose either to continue swallowing event handler errors or not swallow them (personally I'd prefer for the errors to propagate).

Regardless of whether we do any of the above, I'm also curious about warning log statement. Today, I believe this will fire if you have code like:
```
using (var page = ...)
{
    // do stuff
    await page.CloseAsync();
}
```
because CloseAsync gets called twice. This doesn't seem ideal if anyone is looking at the logs. It feels like the intent of this log is to let you know when you've called CloseAsync and the page already died for some other reason. Any thoughts on what the desired behavior is here?

If we don't change any of how `Page.CloseAsync` works, then for test coverage my thought would be to parameterize the current `CloseAsync` tests so that they either call `DisposeAsync` or `CloseAsync` and run both cases. Thoughts?

**Comment by kblok** [comment]
>For Browser, we can implement DisposeAsync as => new ValueTask(CloseAsync()). Because Browser.CloseAsync() caches the returned task, we can conveniently test this by simply asserting that CloseAsync() and DisposeAsync().AsTask() are the same reference. Does that seem like a reasonable testing approach to you or would you prefer to have a set of tests replicating whatever behavior we assert for CloseAsync()?

I wouldn't do much. I think that `DisposeAsync() => CloseAsync();` should be enough. [This test](https://github.com/hardkoded/puppeteer-sharp/blob/7fca5eeeaa9e138af09e24bd98d11d128cac260b/lib/PuppeteerSharp.Tests/PuppeteerTests/PuppeteerLaunchTests.cs#L188) checks that we are closing browsers.
And the CI will end up blowing up if we have process leaks :p
We also write on the console when we open and close browsers, so if during development you see that number grows, you will find a leak.

>We can do the same implementation in theory for Page, but Page.CloseAsync() works a bit differently. Calling CloseAsync() the first time will hit:
>                return Client.Connection.SendAsync("Target.closeTarget", new TargetCloseTargetRequest
                {
                    TargetId = Target.TargetId
                }).ContinueWith(task => Target.CloseTask);
I'm not sure what to make of the ContinueWith, since it is returning a Task<Task> but not calling Unwrap so we don't observe or wait on the inner CloseTask.
The second time CloseAsync() is called, we hit:
_logger.LogWarning("Protocol error: Connection closed. Most likely the page has been closed.");
return _closeCompletedTcs.Task;
This is giving back a different task whose purpose seems to be ensuring that (1) any exceptions hit by firing page close events are swallowed and (2) any continuations will run asynchronously. One strange thing about this is that consecutive calls to close can return tasks with (temporarily) inconsistent completion state.

I see what you mention here, good point.

>I'm wondering whether it makes sense to adjust how Page.CloseAsync() works as part of this work. It seems like it would be safe to always return _closeCompletedTcs.Task. Another option would be to change the Page constructor so that the continuation it creates on Target.CloseTask (which fires the close events) is saved off. We could pass TaskContinuationOptions.RunContinuationsAsynchronously to that continuation to preserve the behavior we were getting from the TCS, and we could choose either to continue swallowing event handler errors or not swallow them (personally I'd prefer for the errors to propagate).

If there's any refactor on the close function the way to go would be [porting how puppeteer performs the close](https://github.com/puppeteer/puppeteer/blame/master/src/Page.js#L1038). It's interesting how we are out of sync there.

>Regardless of whether we do any of the above, I'm also curious about warning log statement. Today, I >believe this will fire if you have code like:
>using (var page = ...)
>{
>    // do stuff
 >   await page.CloseAsync();
>}
>because CloseAsync gets called twice. This doesn't seem ideal if anyone is looking at the logs. It feels like the intent of this log is to let you know when you've called CloseAsync and the page already died for some other reason. Any thoughts on what the desired behavior is here?

`Page` is `IDisposable` just in order to be "developer friendly", it doesn't need to be disposable. You could just call `CloseAsync` and that's it. The using block is just a C# flavor. A developer could tell me that's a bad use of `IDisposable` and I would agree.
In other words, if you use `using` and `CloseAsync` you (not you the developer :p) are missing the point.

>If we don't change any of how Page.CloseAsync works, then for test coverage my thought would be to parameterize the current CloseAsync tests so that they either call DisposeAsync or CloseAsync and run both cases. Thoughts?

We do have [tests to check the `CloseAsync` behavior](https://github.com/hardkoded/puppeteer-sharp/blob/6f813bdc243a74c1a77cddb1e9c440a927c19413/lib/PuppeteerSharp.Tests/PageTests/CloseTests.cs#L28), but we don't have any tests for the `using` block. I think it would be nice to have one.

I think all your points are valid, but scope wise, a `DisposeAsync() => CloseAsync()` should do the trick for both `Browser` and `Page` and we would be super helpful for developers using .netstandar2.1

(feel free to push back if you disagree).

**Comment by madelson** [comment]
Thanks again for the quick response @kblok. Following your suggestion I went for a minimalist changeset (see linked PR).

**Comment by kblok** [comment]
V2.0.3 shipped @madelson !
